spec/validations/requests_spec.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+def request_function(style, address)
+  case style
+  when 'ZAF'
+    "function request() { client.request('#{address}') }"
+  when 'jQuery'
+    "function request() { $.get('#{address}') }"
+  when 'XMLHttpRequest'
+    "function request() { xhr.open('get', '#{address}') }"
+  when 'fetch'
+    "function request() { fetch('#{address}') }"
+  end
+end
+
+request_function_styles = ['ZAF', 'jQuery', 'XMLHttpRequest', 'fetch']
+
+shared_examples 'an insecure request' do |function_style|
+  let(:markup) { request_function(function_style, 'http://foo.com') }
+
+  it "and raise a warning inside #{function_style} style requests" do
+    # TODO: update with translated string
+    warning = 'Possible insecure HTTP request to http://foo.com in assets/app.js. Consider instead using the HTTPS protocol.'
+    errors = subject.call(package)
+    expect(package.warnings).to include(warning)
+    expect(errors).to be_empty
+  end
+end
+
+blocked_requests = [
+  {
+    type: 'private',
+    example_address: 'https://192.168.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to private network address 192.168.0.1 in assets/app.js'
+  },
+  {
+    type: 'loopback',
+    example_address: 'https://127.0.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to a loopback network address 127.0.0.1 in assets/app.js'
+  },
+  { type: 'link-local',
+    example_address: 'https://169.254.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to a link-local network address 169.254.0.1 in assets/app.js'
+  }
+]
+
+blocked_requests.each do |blocked_request|
+  shared_examples "a blocked #{blocked_request[:type]} request" do |function_style|
```

Comment:
This makes me a bit uncomfortable because of the way it forces the test to rely on dynamically-generated magic strings that have to be kept in sync (e.g. modifying a request style means updating the object above, the array of request styles, and the middle of the test description string). I don't have any good suggestions here and I agree that it's better than all the duplication that would be involved in avoiding shared examples entirely.

What if we just had the one `a blocked request` shared example and passed in the type/markup/error as params?

Reply:
Admittedly I think this is a bit gross as well - I scratched my head about this for a long time and eventually ended up here trying to make the assertion readable, e.g. `ZendeskAppsSupport::Validations::Requests to 127.0.0.0/8 range addresses behaves like a blocked loopback request and error inside XMLHttpRequest style request calls`. `it_behaves_like` is really awkward here.

I think your suggestion makes sense... I'll give that a go